lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -243,6 +350,85 @@ void FIRRTLModuleLowering::runOnOperation() {
 }
 
+void FIRRTLModuleLowering::lowerMemoryDecls(const SmallVector<FirMemory> &mems,
+                                            Block *top,
+                                            CircuitLoweringState &state) {
+  auto b =
+      ImplicitLocOpBuilder::atBlockBegin(UnknownLoc::get(&getContext()), top);
+  std::array<StringRef, 8> schemaFields = {
+      "depth",       "numReadPorts", "numWritePorts", "numReadWritePorts",
+      "readLatency", "writeLatency", "width",         "readUnderWrite"};
+  auto schemaFieldsAttr = b.getStrArrayAttr(schemaFields);
+  auto schema = b.create<rtl::RTLGeneratorSchemaOp>(
+      "FIRRTLMem", "FIRRTL_Memory", schemaFieldsAttr);
+  auto memorySchema = b.getSymbolRefAttr(schema);
+
+  Type b1Type = IntegerType::get(&getContext(), 1);
+
+  for (auto &mem : mems) {
+    SmallVector<rtl::ModulePortInfo> ports;
+    size_t inputPin = 0;
+    size_t outputPin = 0;
+
+    auto makePortCommon = [&](StringRef prefix, Twine i, Type bAddrType) {
+      ports.push_back({b.getStringAttr((prefix + "_clock_" + i).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr((prefix + "_en_" + i).str()), rtl::INPUT,
+                       b1Type, inputPin++});
+      ports.push_back({b.getStringAttr((prefix + "_addr_" + i).str()),
+                       rtl::INPUT, bAddrType, inputPin++});
+    };
+
+    Type bDataType =
+        IntegerType::get(&getContext(), std::max((size_t)1, mem.dataWidth));
+
+    Type bAddrType = IntegerType::get(
+        &getContext(), std::max(1U, llvm::Log2_64_Ceil(mem.depth)));
+
+    for (size_t i = 0; i < mem.numReadPorts; ++i) {
+      makePortCommon("ro", Twine(i), bAddrType);
+      ports.push_back({b.getStringAttr(("ro_data_" + Twine(i)).str()),
+                       rtl::OUTPUT, bDataType, outputPin++});
+    }
+    for (size_t i = 0; i < mem.numReadWritePorts; ++i) {
+      makePortCommon("rw", Twine(i), bAddrType);
+      ports.push_back({b.getStringAttr(("rw_wmode_" + Twine(i)).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr(("rw_wmask_" + Twine(i)).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr(("rw_wdata_" + Twine(i)).str()),
+                       rtl::INPUT, bDataType, inputPin++});
+      ports.push_back({b.getStringAttr(("rw_rdata_" + Twine(i)).str()),
+                       rtl::OUTPUT, bDataType, outputPin++});
+    }
+
+    for (size_t i = 0; i < mem.numWritePorts; ++i) {
+      makePortCommon("wo", Twine(i), bAddrType);
+      ports.push_back({b.getStringAttr(("wo_mask_" + Twine(i)).str()),
+                       rtl::INPUT, b1Type, inputPin++});
+      ports.push_back({b.getStringAttr(("wo_data_" + Twine(i)).str()),
+                       rtl::INPUT, bDataType, inputPin++});
+    }
+    std::array<NamedAttribute, 8> genAttrs = {
+        b.getNamedAttr("depth", b.getI64IntegerAttr(mem.depth)),
```

Comment:
This looks like it should be a helper method on FirMemory, to return all the attrs etc.